### PR TITLE
Oauth2 supports authenticated introspection request

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Support client_id,client_secret as a way to authenticated token introspection requests

# Context

Pre https://tools.ietf.org/html/rfc6749#section-2.3.1 this is one technique for adding authentication to requests.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 1]
```diff
@@ -756,6 +778,121 @@ message ExtAuthConfig {
         // AccessTokenValidation access_token_validation = 8;
     }
 
+    message AccessTokenValidationConfig {
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:781]: Since the gloo api and extauth api have now changed, we need to explicitly define the ext auth type.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 2]
```diff
@@ -769,7 +906,16 @@ message ExtAuthConfig {
             // - form-encoded body parameter. recommended, more likely to appear. e.g.: Authorization: Bearer mytoken123
             // - URI query parameter e.g. access_token=mytoken123
             // - and (preferably) secure cookies
-            AccessTokenValidation access_token_validation = 2;
+            // Deprecated: Prefer AccessTokenValidationConfig
+            AccessTokenValidation access_token_validation = 2 [deprecated = true];
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:910]: We build this field internally, should I remove it, and reserve the field?

File: changelog/v1.7.0-beta28/access-token-validation-client-details.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+changelog:
+  - type: NEW_FEATURE
```
Comment (marcogschmidt) [inline changelog/v1.7.0-beta28/access-token-validation-client-details.yaml:2]: Not sure we want to mark this as a new feature here as:
- it will not be implemented when we release and
- this is really an enterprise only feature
I think we can add this changelog entry when we release enterprise.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 4]
```diff
@@ -484,16 +484,38 @@ message AccessTokenValidation {
...
+    // Defines how (opaque) access tokens, received from the oauth authorization endpoint, are validated
+    // [OAuth2.0 Token Introspection](https://tools.ietf.org/html/rfc7662)
+    message IntrospectionValidation {
         // The URL for the [OAuth2.0 Token Introspection](https://tools.ietf.org/html/rfc7662) endpoint.
         // If provided, the (opaque) access token provided or received from the oauth authorization endpoint
         // will be validated against this endpoint, or locally cached responses for this access token.
         string introspection_url = 1;
 
+        // your client id as registered with the issuer
+        // Optional: Use if the authorization server requires client credentials in the request
```
Comment (marcogschmidt) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:496]: Nit: I would rephrase this a bit as "client credentials" is also the name of an OAuth flow:
```suggestion
        // Your client id as registered with the issuer.
        // Optional: Use if the token introspection endpoint requires client authentication.
```

Comment (marcogschmidt) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:496]: Same for the field below.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 5]
```diff
@@ -484,16 +484,38 @@ message AccessTokenValidation {
         string issuer = 3;
     }
 
-    oneof validation_type {
+    // Defines how (opaque) access tokens, received from the oauth authorization endpoint, are validated
+    // [OAuth2.0 Token Introspection](https://tools.ietf.org/html/rfc7662)
```
Comment (marcogschmidt) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:488]: Can you also document how the optional credentials are used here, e.g.:
- both are required, otherwise the config will be rejected (if that is the case)
- they will encoded in a basic auth header

**Review by marcogschmidt — APPROVED** [review]
Looks good!
